libraries/core_libs/network/include/network/tarcap/shared_states/pbft_syncing_state.hpp

```diff
@@ -73,13 +98,34 @@ class PbftSyncingState {
   // Number of seconds needed for ongoing syncing to be declared as inactive
   static constexpr std::chrono::seconds kSyncingInactivityThreshold{60};
 
+  // Number of milliseconds for peer asking sync period timeout
+  // corner case restart/rebuild DB may ask previous periods again
+  const std::chrono::milliseconds kPeerAskingPeriodTimeout{1000};
+
   // What time was received last syncing packet
   std::chrono::steady_clock::time_point last_received_sync_packet_time_{std::chrono::steady_clock::now()};
   mutable std::shared_mutex time_mutex_;
 
   // Peer that the node is syncing with
   std::shared_ptr<TaraxaPeer> peer_;
   mutable std::shared_mutex peer_mutex_;
+
+  /**
+   * @brief Table for peer asking sync period
+   *        Table key: peer node ID, value: a pair
+   *        Pair key: peer last asking request period, pair value: peer last asking request time point
+   */
+  std::unordered_map<dev::p2p::NodeID, std::pair<uint64_t, std::chrono::steady_clock::time_point>>
+      asking_period_peers_table_;
+  mutable boost::shared_mutex asking_period_mutex_;
+
+  /**
+   * @brief Table for incoming last PBFT blocks period for syncing per each peer.
+   *        Key: peer node ID
+   *        Value: last PBFT block period number
+   */
+  std::unordered_map<dev::p2p::NodeID, uint64_t> incoming_blocks_period_peers_table_;
+  mutable boost::shared_mutex incoming_blocks_mutex_;
```

Comment:
I do not think we need this table, as you can sync only from one node = one node id

Reply:
so you should use existing `std::shared_ptr<TaraxaPeer> peer_` and maybe just add period and use existing mutex too

Reply:
In the same example:
1. A node1 first sync with node2, node2 send PBFT blocks B1, B2, B3 to node1
2. Node1 received B1 and B2. But B3 take a long time to deliver(maybe because network delay or something else caused B3 cannot deliver on time)
3. Node1 switch to sync with node3, node3 send B3, B4 to node1
4. Node1 receive B3, B4 from node3
5. node1 receive B3 from node2

I think we need save all peers that the node has synced with from/to

Reply:
ok, then you need to clean this cache, or create expiration map here

Reply:
clean the cache when mark as malicious